Changes to support preview.aptible.com

fancyremarker opened the pull request:
Currently, we rely on Contentful Webhooks to push to Travis CI in order to be able to (a) preview Contentful changes on staging before publishing, and (b) publish Contentful changes without explicitly running a deploy. While this is [Contentful's recommended approach](https://www.contentful.com/developers/docs/ruby/tutorials/automated-rebuild-and-deploy-with-circleci-and-webhooks/), it's got a couple downsides:

1. Changes are slow to be able to preview on staging. The Travis CI build takes about 5 minutes on average, and then CloudFront caching means they may take up to an extra 5 minutes to appear on staging/production. So, it can sometimes take up to 10 minutes for a Contentful change to be visible on staging.
2. The Contentful webhook to Travis CI often fails (429 Too Many Requests), so sometimes (actually, > 30% of the time) changes don't even make it to staging/production.

My proposed solution is:

* Get rid of Contentful webhooks entirely
* Use [preview.aptible.com](https://preview.aptible.com) as a new default place for previewing Contentful changes: This is the subject of this PR, which allows us to host preview.aptible.com on Aptible Enclave, pulling in new content from Contentful once per minute on a cron. Because we're bypassing CloudFront, we also avoid the cache invalidation latency. So Contentful changes should be visible on preview.aptible.com within a minute. We can bring this latency down even further in the future, by adding a webhook listener to preview.aptible.com which can be pinged by Contentful to trigger a new pull from Contentful.
* Deploy to www.aptible.com and www.aptible-staging.com every 15 minutes, via a separate cron app. This means that published changes may take up to 20 minutes to appear on production, but I believe that's an acceptable price to pay for predictability.

This is set up on [preview.aptible.com](https://preview.aptible.com) if you'd like to test it out.

cc: @gib @henryhund

henryhund commented:
This is 100% an upgrade to the current situation. I've struggled with
managing deploys from within Travis and waiting for the deploys to
complete. This is especially painful when experimenting with formatting
within a post.

Will preview.aptible.com pull from the contentful preview API every minute?

I didn't know you were working on this, but I'm grateful for the change and
very much appreciate you proactively improving this. Thank you!


From: Frank Macreery <notifications@github.com> <notifications@github.com>
Reply: aptible/www.aptible.com
<reply@reply.github.com>
<reply@reply.github.com>
Date: June 16, 2017 at 7:43:17 PM
To: aptible/www.aptible.com <www.aptible.com@noreply.github.com>
<www.aptible.com@noreply.github.com>
CC: Henry Hund <henryhund@gmail.com> <henryhund@gmail.com>, Mention
<mention@noreply.github.com> <mention@noreply.github.com>
Subject:  [aptible/www.aptible.com] Changes to support preview.aptible.com
(#494)

Currently, we rely on Contentful Webhooks to push to Travis CI in order to
> be able to (a) preview Contentful changes on staging before publishing, and
> (b) publish Contentful changes without explicitly running a deploy. While
> this is Contentful's recommended approach
> <https://www.contentful.com/developers/docs/ruby/tutorials/automated-rebuild-and-deploy-with-circleci-and-webhooks/>,
> it's got a couple downsides:
>
>    1. Changes are slow to be able to preview on staging. The Travis CI
>    build takes about 5 minutes on average, and then CloudFront caching means
>    they may take up to an extra 5 minutes to appear on staging/production. So,
>    it can sometimes take up to 10 minutes for a Contentful change to be
>    visible on staging.
>    2. The Contentful webhook to Travis CI often fails (429 Too Many
>    Requests), so sometimes (actually, > 30% of the time) changes don't even
>    make it to staging/production.
>
> My proposed solution is:
>
>    - Get rid of Contentful webhooks entirely
>    - Use preview.aptible.com as a new default place for previewing
>    Contentful changes: This is the subject of this PR, which allows us to host
>    preview.aptible.com on Aptible Enclave, pulling in new content from
>    Contentful once per minute on a cron. Because we're bypassing CloudFront,
>    we also avoid the cache invalidation latency. So Contentful changes should
>    be visible on preview.aptible.com within a minute. We can bring this
>    latency down even further in the future, by adding a webhook listener to
>    preview.aptible.com which can be pinged by Contentful to trigger a new
>    pull from Contentful.
>    - Deploy to www.aptible.com and www.aptible-staging.com every 15
>    minutes, via a separate cron app. This means that published changes may
>    take up to 20 minutes to appear on production, but I believe that's an
>    acceptable price to pay for predictability.
>
> This is set up on preview.aptible.com if you'd like to test it out.
>
> cc: @gib <https://github.com/gib> @henryhund
> <https://github.com/henryhund>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/aptible/www.aptible.com/pull/494
> Commit Summary
>
>    - Changes to support preview.aptible.com
>
> File Changes
>
>    - *A* Dockerfile
>    <https://github.com/aptible/www.aptible.com/pull/494/files#diff-0> (20)
>    - *M* Gemfile
>    <https://github.com/aptible/www.aptible.com/pull/494/files#diff-1> (2)
>    - *M* Gemfile.lock
>    <https://github.com/aptible/www.aptible.com/pull/494/files#diff-2> (4)
>    - *A* schedule.rb
>    <https://github.com/aptible/www.aptible.com/pull/494/files#diff-3> (9)
>    - *A* script/aptible-cmd.sh
>    <https://github.com/aptible/www.aptible.com/pull/494/files#diff-4> (12)
>
> Patch Links:
>
>    - https://github.com/aptible/www.aptible.com/pull/494.patch
>    - https://github.com/aptible/www.aptible.com/pull/494.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aptible/www.aptible.com/pull/494>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADWZgRUcnR2Rj7pW3-gyR0MQ2FPNMc0Iks5sExMVgaJpZM4N9DjI>
> .
>

fancyremarker commented:
> Will preview.aptible.com pull from the contentful preview API every minute?

Yup, exactly. Eventually it would be nice to make this happen more instantaneously, via a webhook listener in the Middleman server itself. I'll take a look into what that would involve.

@gib and I were happy to help with this, since the current Travis CI webhook integration isn't great for us either. He and I divided and conquered the work to make this happen and he took the bigger chunk (currently WIP in #492).

gib commented:
Thank you @fancyremarker. This is really exciting. The predictability is a huge win and better all around with the preview. 🎉

fancyremarker commented:
As mentioned [in Slack](https://aptible.slack.com/archives/C03RK9JBZ/p1497714151934266), one issue with the every-minute cron is that Middleman evidently leaks memory every time it reloads. So, I may look into reloading via cron after all. Will open as a separate PR, for comparison.

fancyremarker commented:
I've added a few improvements to replace the earlier 124bce0, namely:

* Fixed staleness issues @gib and @henry uncovered during a demo (posts deleted from Contentful were not being deleted locally during `rake contentful:pull`, and Middleman wasn't being properly reloaded on file changes, so new posts weren't appearing in proxy-generated index pages)
* Switched to Supercronic (cc: @krallin)
* Updated `rake contentful:pull` to only write to files when they are modified; this should reduce the frequency with which Middleman needs to reload (i.e., it will only reload when there is actually new Contentful content), and should in turn mitigate the RAM leak associated with each Middleman reload

This is up on https://preview.aptible.com and should be ready to merge. Please let me know if anything doesn't behave as expected on https://preview.aptible.com.
